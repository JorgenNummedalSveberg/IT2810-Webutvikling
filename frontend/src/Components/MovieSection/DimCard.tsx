import {Grid, Paper} from "@material-ui/core";
import {Skeleton} from "@material-ui/lab";
import React from "react";

export default function DimCard(props: { classes: any }) {
    return (
        <Grid className={props.classes().gridItem} item>
            <Paper className={props.classes().paper}>
                <Skeleton animation="pulse" variant="rect" width={300} height={450}/>
                <div className={props.classes().details}>
                    <Skeleton animation="wave" width={180} height={40} style={{margin: 10}}/>
                    <Skeleton animation="wave" width={200} height={2} style={{margin: 10}}/>
                    <Skeleton animation="wave" width={170} height={40} style={{margin: 10}}/>
                    <Skeleton animation="wave" width={130} height={40} style={{margin: 10}}/>
                </div>
            </Paper>
        </Grid>
    )
}