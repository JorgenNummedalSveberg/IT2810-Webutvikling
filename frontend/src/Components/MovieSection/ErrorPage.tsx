import {Divider, Paper, Typography} from "@material-ui/core";
import React from "react";

export default function errorPage(props: {
    classes: {errorPage: string, details: string; gridItem: string; title: string; description: string; bottomInfo: string}}){
    return (
        <div className={props.classes.errorPage}>
            <Paper className={`${props.classes.details} ${props.classes.gridItem}`} elevation={5}>
                <div className={props.classes.title}>
                    <Typography color='error' variant="h5" component="h2">
                        No movies
                    </Typography>
                    <Divider/>
                </div>
                <div className={props.classes.description}>
                    This might be because:
                </div>
                <div className={props.classes.bottomInfo}>
                    <ul>
                        <li><Typography color='error'>You may not be on the NTNU network or your VPN is off</Typography>
                        </li>
                        <li><Typography color='error'>We do not have the movie you're looking for</Typography></li>
                    </ul>
                </div>
                <a href={'https://www.youtube.com/watch?v=oHg5SJYRHA0'}>Maybe this can help</a>
            </Paper>
        </div>
    )
}