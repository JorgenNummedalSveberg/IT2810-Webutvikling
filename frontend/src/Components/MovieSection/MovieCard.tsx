import {Movie} from "../../types/Movie";
import {useDispatch} from "react-redux";
import {setPopup, showPopup} from "../../actions";
import {
    ButtonBase,
    Card,
    CardContent,
    CardHeader,
    CardMedia,
    Divider,
    Grid,
    Paper,
    useMediaQuery
} from "@material-ui/core";
import {Skeleton} from "@material-ui/lab";
import HourglassFullIcon from '@material-ui/icons/HourglassFull';
import ImdbIcon from "../Shared/ImdbIcon";
import React from "react";
import {makeStyles} from "@material-ui/styles";

// Komponent for å vise frem en film i et kort
function MovieCard(props: { movie: Movie }) {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Åpner opp Popup.
    function handleClick() {
        dispatch(setPopup(props.movie));
        dispatch(showPopup(true));
    }

    const classes = makeStyles({
        card: {height: "100%", width: '100%', backgroundColor: '#464646'},
        gridItem: {flexGrow:1, flexBasis: 1, maxWidth: '600px', width: '600px'},
        paperButton: {height: "100%", width: '100%'},
        paper: {backgroundColor: '#E85A4F', height: "100%", width: '100%', display: 'flex', flexDirection: 'row'},
        poster: {maxWidth: '500px', minWidth: '250px'},
        details: {padding: '10px', display: 'flex', flexDirection: 'column'},
        title: {flexGrow: 1},
        description: {flexGrow: 4, color: 'white', textAlign: 'left'},
        bottomInfo: {
            margin: '10px',
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
            flexGrow: 1,
            '& *': {margin: '5px'}
        },
        duration: {color: '#e5dfca', display: 'flex', alignItems: 'center'},
        noMargin: {
            margin: 0
        }
    })

    return (
        <Grid className={classes().gridItem} item>
            <ButtonBase className={classes().paperButton} onClick={handleClick}>
                <Paper className={classes().paper} elevation={5}>
                    <img className={classes().poster} alt='Could not display movie poster' width='100%' src={props.movie.posterurl}/>
                    <div className={classes().details}>
                        <div className={classes().title}>
                            <h1>{props.movie.title}</h1>
                            <Divider/>
                        </div>
                        <div className={classes().description}>
                            <h3>Year: {props.movie.year}</h3>
                            <h3>Genres: {props.movie.genres.join(', ')}</h3>
                        </div>
                        <div className={classes().bottomInfo}>
                            <div className={classes().duration}>
                                <HourglassFullIcon className={classes().noMargin} fontSize={"large"}/>
                                <p className={classes().noMargin}>{parseTime(props.movie.duration)}</p>
                            </div>
                            <Divider orientation='vertical'/>
                            <ImdbIcon rating={props.movie.imdbRating} height={35}/>
                        </div>
                    </div>
                </Paper>
            </ButtonBase>
        </Grid>

    )
}

// Blanke kort for når nettsiden laster inn filmene
export function DimCard() {
    return (
        <Grid item>
            <Card style={{height: "100%", width: '100%', backgroundColor: '#464646'}}>
                <Skeleton style={{backgroundColor: '#222222'}} variant="rect" width={"100%"} height={300} />
                <Skeleton animation="wave" width={340} height={40} style={{ margin: 10 }} />
                <Skeleton animation="wave" width={340} height={40} style={{ margin: 10 }} />
            </Card>
        </Grid>
    )
}

// Tar inn tid i formatet på databasen og gjør det om til presentabel string, eller rein minuttverdi
function parseTime(time: number): string {
    let minutes = time;
    let hours = 0;
    while (minutes - 60 > 0) {
        minutes -= 60;
        hours++;
    }
    return hours + 'h' + minutes + 'm';
}

export default MovieCard