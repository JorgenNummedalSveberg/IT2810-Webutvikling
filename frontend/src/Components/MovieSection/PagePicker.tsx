import {setPage} from "../../actions";
import {Pagination} from "@material-ui/lab";
import React from "react";

export default function
    PagePicker(props: {
    dispatch: (f: { payload: any; type: string }) => void,
    page: number,
    pages: number,
    refresh: (number: number) => void
}) {
    return (
        <Pagination
            color={'primary'}
            size="large"
            onChange={(e: object, page: number) => {
                props.dispatch(setPage(page - 1));
                props.refresh(page - 1);
            }}
            page={props.page + 1}
            count={props.pages}/>
    )
}