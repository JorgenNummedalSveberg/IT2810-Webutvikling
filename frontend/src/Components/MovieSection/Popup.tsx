import React from 'react';
import {useDispatch, useSelector} from "react-redux";
import {State} from "../../types/State";
import {Button} from "@material-ui/core";
import VisibilityIcon from '@material-ui/icons/Visibility';
import DeleteIcon from '@material-ui/icons/Delete';
import {login} from "../../actions";
import ImdbIcon from "../Shared/ImdbIcon";
import {makeStyles} from "@material-ui/styles";


function Popup(props: { refresh: (page: number) => void }) {

    // Henter state fra redux
    const reduxState = useSelector((state: State) => state);

    // NÃ¸dvendig for redux
    const dispatch = useDispatch();

    let req = {};
    if (!!reduxState.user) {
        req = ({
            method: 'POST',
            body: JSON.stringify({userName: reduxState.user.userName, movieId: reduxState.details.movie._id}),
            headers: {
                'Content-Type': 'application/json'
            }
        })
    }

    // Tar vekk en view og fjerner den fra lista di
    function changeView(remove: boolean) {
        fetch('http://localhost:5000/api/user/' + (remove ? 'removeMovie' : 'addMovie'), req)
            .then(response => {
                if (response.ok) {
                    if (remove) {
                        reduxState.user.movies = reduxState.user.movies.filter(movieId => movieId !== reduxState.details.movie._id);
                    } else {
                        reduxState.user.movies.push(reduxState.details.movie._id);
                    }
                    dispatch(login(reduxState.user));
                    props.refresh(reduxState.page)
                }
            })
            .catch(error => console.log(error));
    }

    const classes = makeStyles({
        root: {
            margin: '0 20px 0 20px',
            display: reduxState.details.show ? 'flex' : 'none',
            flexDirection: 'column',
            alignItems: 'center'
        },
        buttons: {
            display: 'flex',
            flexDirection: 'column',
        }
    })

    return (
        <div className={classes().root}>
            <h1>{reduxState.details.movie.title}</h1>
            <img width='400px' alt="movie poster could not load" src={reduxState.details.movie.posterurl}/>
            <h2>{reduxState.details.movie.year}</h2>
            <ImdbIcon rating={reduxState.details.movie.imdbRating} height={50}/>
            <div className={classes().buttons}>
                {!!reduxState.user ?
                    <Button id='watchButton'
                            variant='contained'
                            disabled={reduxState.user.movies.includes(reduxState.details.movie._id)}
                            onClick={() => changeView(false)}
                            color='primary'
                            endIcon={<VisibilityIcon/>}
                    >Watched</Button>
                    : null}
                {!!reduxState.user && reduxState.user.movies.includes(reduxState.details.movie._id) ?
                    <Button id='removeButton'
                            variant='contained'
                            color='secondary'
                            onClick={() => changeView(true)}
                            endIcon={<DeleteIcon/>}
                    >Remove from my list</Button>
                    :
                    null}
            </div>
            <h3>{reduxState.details.movie.genres.join(", ")}</h3>
            <p>{reduxState.details.movie.storyline}</p>
        </div>
    )
}

export default Popup;