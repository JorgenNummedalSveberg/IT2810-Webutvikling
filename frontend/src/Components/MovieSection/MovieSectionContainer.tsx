import React from 'react';
import {useMediaQuery, useTheme} from '@material-ui/core'
import {useDispatch, useSelector} from "react-redux";
import {State} from "../../types/State";
import {setPopup, showPopup} from "../../actions";
import MovieCard from "./MovieCard";
import DimCard from "./DimCard";
import {makeStyles} from "@material-ui/styles";
import {Movie} from "../../types/Movie";
import MovieSection from "./MovieSection";
import {parseTime} from "../../movieService";

// Komponent som viser frem alle filmene i en responsiv grid
function MovieSectionContainer(props: { refresh: (number: number) => void, error: boolean }) {

    // Henter inn redux state
    const state = useSelector((state: State) => state);

    const theme = useTheme();
    const classes = makeStyles({
        root: {marginLeft: useMediaQuery('(max-width: 1400px)').valueOf() ? '' : '500px'},
        main: {display: 'flex', flexDirection: 'column', alignItems: 'center', padding: '20px', overflowY: 'auto',},
        pagination: {color: theme.palette.getContrastText('#445585')},
        movieGrid: {width: '100%', margin: '20px',},
        popup: {
            backgroundColor: theme.palette.primary.light,
            textAlign: 'center',
            width: useMediaQuery('(max-width: 1400px)').valueOf() ? '100%' : '30vw',
            height: '100%',
            paddingTop: '5%',
            paddingBottom: '5%'
        },
        errorPage: {display: props.error ? 'flex' : 'none', justifyContent: 'center', padding: '50px'},
        moviePage: {display: props.error ? 'none' : 'initial'},
        card: {height: "100%", width: '100%'},
        gridItem: {flexGrow: 1, flexBasis: 1, maxWidth: '600px', width: '600px'},
        paperButton: {height: "100%", width: '100%'},
        paper: {
            backgroundColor: theme.palette.primary.light,
            height: "100%",
            width: '100%',
            display: 'flex',
            flexDirection: 'row'
        },
        poster: {maxWidth: '500px', minWidth: '250px'},
        details: {padding: '10px', display: 'flex', flexDirection: 'column'},
        title: {flexGrow: 1},
        description: {
            flexGrow: 4,
            color: theme.palette.getContrastText(theme.palette.primary.light),
            textAlign: 'left'
        },
        bottomInfo: {
            margin: '10px',
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
            flexGrow: 1,
            '& *': {margin: '5px'}
        },
        duration: {
            color: theme.palette.getContrastText(theme.palette.primary.light),
            display: 'flex',
            alignItems: 'center'
        },
        noMargin: {margin: 0}
    })

    // Lager en liste med sorte kort som placeholder mens filmene laster
    const dimList = () => {
        const list = [];
        for (let i = 0; i < 24; i++) {
            list.push(<DimCard classes={classes} key={i}/>);
        }
        return list;
    }


    // Nødvendig for redux
    const dispatch = useDispatch();

    // Åpner opp Popup.
    function handleClick(movie: Movie) {
        dispatch(setPopup(movie));
        dispatch(showPopup(true));
    }

    // Lager en liste av alle MovieCards som skal med i Griden
    let movieCards: any[] = [];
    const movieList = state.movieCache.filter(movie => state.indexList.includes(movie._id))
        .sort((a, b) => state.indexList.indexOf(a._id) - state.indexList.indexOf(b._id))
    if (state.indexList.length > 0) {
        movieCards = movieList.map((movie: Movie, index: number) => {
            return (
                <MovieCard handleClick={handleClick} classes={classes} movie={movie} key={index}
                           duration={parseTime(movie.duration)}/>
            )
        })
    }

    return (
        <MovieSection
            myMovies={state.filter.myMovies}
            page={state.page}
            pages={state.pages}
            classes={classes}
            dispatch={useDispatch()}
            movieCards={movieCards.length === state.indexList.length && movieCards.length > 0 ? movieCards : dimList()}
            popupShow={state.details.show}
            refresh={props.refresh}/>
    )
}

export default MovieSectionContainer;