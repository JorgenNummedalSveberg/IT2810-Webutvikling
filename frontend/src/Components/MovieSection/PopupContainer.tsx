import React from 'react';
import {useDispatch, useSelector} from "react-redux";
import {State} from "../../types/State";
import {login} from "../../actions";
import {makeStyles} from "@material-ui/styles";
import Popup from "./Popup";


function PopupContainer(props: { refresh: (page: number) => void }) {

    // Henter state fra redux
    const reduxState = useSelector((state: State) => state);

    // NÃ¸dvendig for redux
    const dispatch = useDispatch();

    let req = {};
    if (!!reduxState.user) {
        req = ({
            method: 'POST',
            body: JSON.stringify({userName: reduxState.user.userName, movieId: reduxState.details.movie._id}),
            headers: {
                'Content-Type': 'application/json'
            }
        })
    }

    // Tar vekk en view og fjerner den fra lista di
    function changeView(remove: boolean) {
        fetch('http://localhost:5000/api/user/' + (remove ? 'removeMovie' : 'addMovie'), req)
            .then(response => {
                if (response.ok) {
                    if (remove) {
                        reduxState.user.movies = reduxState.user.movies.filter(movieId => movieId !== reduxState.details.movie._id);
                    } else {
                        reduxState.user.movies.push(reduxState.details.movie._id);
                    }
                    dispatch(login(reduxState.user));
                    props.refresh(reduxState.page)
                }
            })
            .catch(error => console.log(error));
    }

    const classes = makeStyles({
        root: {
            margin: '0 20px 0 20px',
            display: reduxState.details.show ? 'flex' : 'none',
            flexDirection: 'column',
            alignItems: 'center'
        },
        buttons: {
            display: 'flex',
            flexDirection: 'column',
        }
    })

    return (
        <Popup
            movie={reduxState.details.movie}
            changeView={changeView}
            user={reduxState.user}
            classes={classes()}
            refresh={props.refresh}/>
    )
}

export default PopupContainer;