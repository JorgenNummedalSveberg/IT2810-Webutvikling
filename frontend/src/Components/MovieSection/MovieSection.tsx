import {Button, Drawer, Grid} from "@material-ui/core";
import {showPopup} from "../../actions";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import Popup from "./Popup";
import React from "react";
import ErrorPage from "./ErrorPage";
import PagePicker from "./PagePicker";
import PopupContainer from "./PopupContainer";

export default function MovieSection(props: {
    page: number,
    pages: number
    classes: any,
    dispatch: (f: { payload: any; type: string }) => void,
    popupShow: boolean,
    refresh: (number: number) => void,
    movieCards: JSX.Element[]}) {
    return (
        <div className={props.classes().root}>
            <ErrorPage classes={
                {errorPage: props.classes().errorPage,
                    details: props.classes().details,
                    gridItem: props.classes().gridItem,
                    title: props.classes().title,
                    description: props.classes().description,
                    bottomInfo: props.classes().bottomInfo}}/>
            <div className={props.classes().moviePage}>
                <Drawer anchor={'right'} open={props.popupShow} onClose={() => props.dispatch(showPopup(false))}>
                    <Button startIcon={<ArrowBackIcon/>} onClick={() => props.dispatch(showPopup(false))}>Close</Button>
                    <div className={props.classes().popup}>
                        <PopupContainer refresh={props.refresh}/>
                    </div>
                </Drawer>
                <div className={props.classes().main}>
                    <PagePicker refresh={props.refresh} dispatch={props.dispatch} page={props.page} pages={props.pages}/>
                    <Grid
                        className={props.classes().movieGrid}
                        container
                        justify="center"
                        alignItems="stretch"
                        spacing={4}
                    >
                        {props.movieCards}
                    </Grid>
                    <PagePicker refresh={props.refresh} dispatch={props.dispatch} page={props.page} pages={props.pages}/>
                </div>
            </div>
        </div>
    )
}