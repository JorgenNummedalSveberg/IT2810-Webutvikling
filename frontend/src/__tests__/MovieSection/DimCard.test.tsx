import {render} from "@testing-library/react";
import React from "react";
import {DimCard} from "../../Components/MovieSection/MovieCard";
import {mock} from "./MovieSection.test";


describe("A snapshot test for the ImdbIcon", () => {
    test("renders, and matches with snapshot", () => {
        const {container} = render(<DimCard classes={mock}/>);
        expect(container).toMatchSnapshot();
    })
})

