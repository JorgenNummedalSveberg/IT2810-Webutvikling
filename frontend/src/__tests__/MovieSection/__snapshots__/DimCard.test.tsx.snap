// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A snapshot test for the ImdbIcon renders, and matches with snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-gridItem-2 MuiGrid-item"
  >
    <div
      class="MuiPaper-root makeStyles-paper-4 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="width: 300px; height: 450px;"
      />
      <div
        class="makeStyles-details-6"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 180px; height: 40px; margin: 10px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 200px; height: 2px; margin: 10px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 170px; height: 40px; margin: 10px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 130px; height: 40px; margin: 10px;"
        />
      </div>
    </div>
  </div>
</div>
`;
